apps/openmw/mwmechanics/aicombat.cpp

```diff
@@ -326,95 +284,66 @@ namespace MWMechanics
         if (currentAction.get())
             currentAction->getCombatRange(rangeAttack, rangeFollow);
 
-        // FIXME: consider moving this stuff to ActionWeapon::getCombatRange
         const ESM::Weapon *weapon = NULL;
-        MWMechanics::WeaponType weaptype = WeapType_None;
-        float weapRange = 1.0f;
+        MWMechanics::WeaponType weapType = MWMechanics::WeaponType::WeapType_None;
+        float spellSpeed = 1.f;
 
-        // Get weapon characteristics
-        if (actorClass.hasInventoryStore(actor))
+        if (dynamic_cast<ActionWeapon*>(currentAction.get()))
         {
-            // TODO: Check equipped weapon and equip a different one if we can't attack with it
-            // (e.g. no ammunition, or wrong type of ammunition equipped, etc. autoEquip is not very smart in this regard))
+            weapon = dynamic_cast<ActionWeapon*>(currentAction.get())->getWeapon();
 
-            //Get weapon speed and range
-            MWWorld::ContainerStoreIterator weaponSlot =
-                MWMechanics::getActiveWeapon(actorClass.getCreatureStats(actor), actorClass.getInventoryStore(actor), &weaptype);
+            if (weapon != NULL)
+                weapType = getWeaponType(static_cast<ESM::Weapon::Type>(weapon->mData.mType));
+            else
+                weapType = WeapType_HandToHand;
 
-            if (weaptype == WeapType_HandToHand)
-            {
-                static float fHandToHandReach =
-                    world->getStore().get<ESM::GameSetting>().find("fHandToHandReach")->getFloat();
-                weapRange = fHandToHandReach;
-            }
-            else if (weaptype != WeapType_PickProbe && weaptype != WeapType_Spell && weaptype != WeapType_None)
+            if ((weapon == NULL && storage.mLastWeapon != "handToHand") 
+                || (weapon != NULL && weapon->mModel != storage.mLastWeapon))
             {
-                // All other WeapTypes are actually weapons, so get<ESM::Weapon> is safe.
-                weapon = weaponSlot->get<ESM::Weapon>()->mBase;
-                weapRange = weapon->mData.mReach;
+                storage.updateWeaponAttackDurations(actor, weapType, weapon == NULL ? 1.f : weapon->mData.mSpeed, ESM::RT_Self);
```

Comment:
We don't need the CharacterController in startCombat. startCombat just contructs the AiPackage, but doesn't execute it.

Btw, now I remember there were some other motivations for having access to the CharacterController.

To balance the combat AI with respect to vanilla MW ( [#1876](https://bugs.openmw.org/issues/1876) ), we need to insert a random delay between the end of the last attack and the start of the next one. As of now the AI doesn't even have sufficient knowledge about the attack process to accurately time that delay. We don't know how long the "follow" part of the attack animation is. The current timing is mostly based on guesswork.

Neither does the AI have knowledge of animation states e.g. knockdown, blocking, ... that would affect the attack timings. These are only known to the CharacterController.

> As I understand this will require some calculations in animation handler class (?), and you going to ask it every frame. 

The animation handler already does that calculation every frame anyway, since it needs to know when to start the next animation. See here: https://github.com/OpenMW/openmw/blob/master/apps/openmw/mwmechanics/character.cpp#L1117